File: juce_port/juce_project/Source/Theory/Scales.cpp
#include "Scales.h"

// Инициализация статических членов
std::vector<int> Scales::major = {0, 2, 4, 5, 7, 9, 11};
std::vector<int> Scales::minor = {0, 2, 3, 5, 7, 8, 10};
std::vector<int> Scales::dorian = {0, 2, 3, 5, 7, 9, 10};
std::vector<int> Scales::phrygian = {0, 1, 3, 5, 7, 8, 10};
std::vector<int> Scales::lydian = {0, 2, 4, 6, 7, 9, 11};
std::vector<int> Scales::mixolydian = {0, 2, 4, 5, 7, 9, 10};
std::vector<int> Scales::locrian = {0, 1, 3, 4, 7, 8, 10};
std::vector<int> Scales::harmonicMinor = {0, 2, 3, 5, 7, 8, 11};
std::vector<int> Scales::melodicMinor = {0, 2, 3, 5, 7, 9, 11};
std::vector<int> Scales::majorPentatonic = {0, 2, 4, 7, 9};
std::vector<int> Scales::minorPentatonic = {0, 3, 5, 7, 10};
std::vector<int> Scales::blues = {0, 3, 5, 6, 7, 10};
std::vector<int> Scales::wholeTone = {0, 2, 4, 6, 8, 10};
std::vector<int> Scales::chromatic = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11};
std::vector<int> Scales::hirajoshi = {0, 2, 3, 7, 8};
std::vector<int> Scales::iwato = {0, 1, 5, 6, 10};
std::vector<int> Scales::hungarianMinor = {0, 2, 3, 6, 7, 8, 11};
std::vector<int> Scales::phrygianDominant = {0, 1, 4, 5, 7, 8, 10};
std::vector<int> Scales::enigmatic = {0, 1, 4, 6, 8, 10, 11};

// Карта соответствия имен ладов и их интервалов
const std::map<juce::String, std::vector<int>> Scales::scaleIntervals = {
    {"Major", major},
    {"Minor", minor},
    {"Dorian", dorian},
    {"Phrygian", phrygian},
    {"Lydian", lydian},
    {"Mixolydian", mixolydian},
    {"Locrian", locrian},
    {"Harmonic Minor", harmonicMinor},
    {"Melodic Minor", melodicMinor},
    {"Major Pentatonic", majorPentatonic},
    {"Minor Pentatonic", minorPentatonic},
    {"Blues", blues},
    {"Whole Tone", wholeTone},
    {"Chromatic", chromatic},
    {"Hirajoshi", hirajoshi},
    {"Iwato", iwato},
    {"Hungarian Minor", hungarianMinor},
    {"Phrygian Dominant", phrygianDominant},
    {"Enigmatic", enigmatic}
};

bool Scales::isNoteInScale(int note, const juce::String& scaleName)
{
    auto it = scaleIntervals.find(scaleName);
    if (it == scaleIntervals.end())
        return false;

    // Получаем ноту относительно октавы (0-11)
    int noteInOctave = note % 12;

    // Проверяем, есть ли нота в интервалах лада
    const auto& intervals = it->second;
    return std::find(intervals.begin(), intervals.end(), noteInOctave) != intervals.end();
}

std::vector<int> Scales::getScaleNotes(int rootNote, const juce::String& scaleName)
{
    return getScaleNotesInRange(rootNote, scaleName, 0, 127);
}

std::vector<int> Scales::getScaleNotesInRange(int rootNote, const juce::String& scaleName, int minNote, int maxNote)
{
    auto it = scaleIntervals.find(scaleName);
    if (it == scaleIntervals.end())
        return {};

    const auto& intervals = it->second;
    if (intervals.empty())
        return {};

    std::vector<int> notes;

    // Генерируем ноты для всех октав
    for (int octave = -1; octave <= 10; ++octave) {  // От C-1 до C10
        for (int interval : intervals) {
            int note = rootNote + interval + (octave * 12);
            if (note >= minNote && note <= maxNote) {
                notes.push_back(note);
            }
        }
    }

    // Сортируем и удаляем дубликаты
    std::sort(notes.begin(), notes.end());
    auto last = std::unique(notes.begin(), notes.end());
    notes.erase(last, notes.end());

    return notes;
}

juce::StringArray Scales::getAvailableScaleNames()
{
    juce::StringArray names;
    for (const auto& pair : scaleIntervals) {
        names.add(pair.first);
    }
    return names;
}

File: stuff/src/theory/scales.py
# Music theory constants and helpers

NOTE_NAMES = ["C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B"]

SCALES = {
    "Major": [0, 2, 4, 5, 7, 9, 11],
    "Minor": [0, 2, 3, 5, 7, 8, 10],
    "Dorian": [0, 2, 3, 5, 7, 9, 10],
    "Phrygian": [0, 1, 3, 5, 7, 8, 10],
    "Lydian": [0, 2, 4, 6, 7, 9, 11],
    "Mixolydian": [0, 2, 4, 5, 7, 9, 10],
    "Locrian": [0, 1, 3, 4, 7, 8, 10],
    "Harmonic Minor": [0, 2, 3, 5, 7, 8, 11],
    "Melodic Minor": [0, 2, 3, 5, 7, 9, 11],
    "Major Pentatonic": [0, 2, 4, 7, 9],
    "Minor Pentatonic": [0, 3, 5, 7, 10],
    "Blues": [0, 3, 5, 6, 7, 10],
    "Whole Tone": [0, 2, 4, 6, 8, 10],
    "Chromatic": list(range(12)),
    "Hirajoshi": [0, 2, 3, 7, 8],
    "Iwato": [0, 1, 5, 6, 10],
    "Hungarian Minor": [0, 2, 3, 6, 7, 8, 11],
    "Phrygian Dominant": [0, 1, 4, 5, 7, 8, 10],
    "Enigmatic": [0, 1, 4, 6, 8, 10, 11],
}

def get_notes_in_scale(root_note: int, scale_intervals: list[int], min_note: int = 0, max_note: int = 127) -> list[int]:
    """
    Generates a list of all MIDI notes for a given scale within a specified range.

    Args:
        root_note: The root note of the scale (0-11).
        scale_intervals: A list of intervals from the root note (e.g., [0, 2, 4, ...]).
        min_note: The minimum MIDI note number for the output.
        max_note: The maximum MIDI note number for the output.

    Returns:
        A sorted list of MIDI note numbers that fall within the scale and range.
    """
    if not scale_intervals:
        return []

    notes = set()
    # Generate notes for all octaves
    for octave in range(-1, 11):  # From C-1 to C10
        for interval in scale_intervals:
            note = root_note + interval + (octave * 12)
            if min_note <= note <= max_note:
                notes.add(note)

    return sorted(list(notes))
